File: assignments/assignment-2-solution/src-csharp/TeaShop/Domain/Query/QueriedInventoryItem.cs
using Assignment2Solution.Domain.Inventory;

namespace Assignment2Solution.Domain.Query;

/// <summary>
///     Represents an inventory item that has been returned from a query, including its display index.
/// </summary>
/// <param name="Index">The display index of the item (1-based).</param>
/// <param name="InventoryItemId">The unique identifier for the inventory item.</param>
/// <param name="Name">The name of the tea.</param>
/// <param name="Price">The price per unit.</param>
/// <param name="Quantity">The available quantity.</param>
/// <param name="StarRating">The star rating of the tea.</param>
public record QueriedInventoryItem(
    int Index,
    Guid InventoryItemId,
    string Name,
    decimal Price,
    int Quantity,
    StarRating StarRating) : InventoryItem(InventoryItemId, Name, Price, Quantity, StarRating)
{
    /// <summary>
    ///     Initializes a new instance of the <see cref="QueriedInventoryItem" /> record based on an existing
    ///     <see cref="InventoryItem" />.
    /// </summary>
    /// <param name="index">The display index of the item.</param>
    /// <param name="item">The inventory item to wrap.</param>
    public QueriedInventoryItem(int index, InventoryItem item)
        : this(index, item.InventoryItemId, item.Name, item.Price, item.Quantity, item.StarRating)
    {
    }
}

File: assignments/assignment-2-solution/src-java/src/main/java/assignment2solution/domain/inventoryquery/QueriedInventoryItem.java
package assignment2solution.domain.inventoryquery;

import assignment2solution.domain.inventory.InventoryItem;
import assignment2solution.domain.inventory.StarRating;

import java.math.BigDecimal;
import java.util.Objects;
import java.util.UUID;

/**
 * Represents an inventory item that has been returned from a query, including its display index.
 */
public final class QueriedInventoryItem extends InventoryItem {
    private final int index;

    public QueriedInventoryItem(int index, UUID inventoryItemId, String name, BigDecimal price, int quantity,
                                StarRating starRating) {
        super(inventoryItemId, name, price, quantity, starRating);
        this.index = index;
    }

    /**
     * Initializes a new instance of the {@link QueriedInventoryItem} class based on an existing
     * {@link InventoryItem}.
     */
    public QueriedInventoryItem(int index, InventoryItem item) {
        this(index,
            Objects.requireNonNull(item, "item").getInventoryItemId(),
            item.getName(),
            item.getPrice(),
            item.getQuantity(),
            item.getStarRating());
    }

    /**
     * Gets the display index of the item (1-based).
     */
    public int getIndex() {
        return index;
    }
}
